Add basic GPU support to distributed autograd.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40312 Add basic GPU support to distributed autograd.**

As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

pritamdamania committed:
Add basic GPU support to distributed autograd.

As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -443,7 +432,7 @@ void Engine::reentrant_thread_init() {
     local_ready_queue = ready_queue_by_index(graph_task->cpu_ready_queue_, graph_task->owner_);
     total_depth = graph_task->reentrant_depth_;
-    thread_main(graph_task, /* reentrant thread*/ true);
+    thread_main(graph_task, /* device_thread */ false);
```
It's a bit confused on the argument changed here, device threads can also have reentrant backwards right?

albanD replied:
I would agree here. Since we have multiple use case for this, it might be clearer to just describe what it does.
Or just remove this flag that is completely redundant with the first arguments :D

pritamdamania87 commented:
Verified the RPC examples are working with this PR.

wanchaol reviewed (approved):
looks good, the `device_thread` naming is confusing as reentrant backwards could happen in the device threads as well.  Maybe call it sth like `spin` or `long_live`?

pritamdamania committed:
Update on "Add basic GPU support to distributed autograd."


As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add basic GPU support to distributed autograd."


As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -282,12 +286,10 @@ void Engine::thread_init(int device, const std::shared_ptr<ReadyQueue>& ready_qu
-  thread_main(graph_task, /* reentrant_thread */ false);
-  // Notify about shutdown
-  {
-    std::unique_lock<std::mutex> lk(non_reentrant_device_thread_mutex_);
```
Was it a bug that this was not properly protected by the if statement?

pritamdamania87 replied:
Not really a bug since we mostly use PyThonEngine and it was taken care of there (since we always pass false here). I just felt it made sense to have this API do the right thing standalone too.

albanD commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -57,6 +57,11 @@ void PythonEngine::thread_init(int device, const std::shared_ptr<ReadyQueue>& re
   pybind11::gil_scoped_release no_gil;
   Engine::thread_init(device, ready_queue, false);
+
+  if (should_increment) {
```
Was it a bug that this was missing?

pritamdamania87 replied:
Well it wasn't really a bug since it kinda worked. PyThonEngine would increment the count and pass false to Engine. Then, engine would always decrement without checking `should_increment`. I just felt it was kinda weird how both the classes had this contract with each other :) This just makes each one consistent on its own.

albanD replied:
Haha, thanks for the explanation.
We do have a growing user-base for libtorch that uses the non-python version so definitely a good fix !

albanD reviewed (commented):
Good catch for the increment/decrement issues !
Just a minor point about argument naming but I think the overall logic is good.

pritamdamania committed:
Update on "Add basic GPU support to distributed autograd."


As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update. LGTM

pritamdamania committed:
Update on "Add basic GPU support to distributed autograd."


As part of https://github.com/pytorch/pytorch/issues/40255, we
realized that GPU support for distributed autograd was broken as part of our
multithreaded autograd change.

To fix this in the short term for 1.6, this PR includes the following changes:

1) Long lived CPU thread in DistEngine to execute GPU->CPU continuations in the
autograd graph.
2) The long lived CPU thread has its own ready_queue and this queue is used for
all GraphTasks created by DistEngine.
3) In thread_main(), the CPU thread cannot exit once the GraphTask is done
processing because of the new CPU thread added in 1).
4) To resolve this, thread_main() now has a parameter `device_thread` instead
of `reentrant_thread`. When device_thread is True, we expect this to be a long
lived device thread that does not exit.
5) When device_thread is False, thread_main is expected to run a GraphTask and
return once done.

Differential Revision: [D22146183](https://our.internmc.facebook.com/intern/diff/D22146183/)

[ghstack-poisoned]

mrshenli commented:
Shall we land this today before branch cut?
